libs/react-helpers/src/lib/generic-data-provider.ts

```diff
@@ -433,29 +451,137 @@ const memoize = <Data, Delta>(
  *
  */
 export function makeDataProvider<QueryData, Data, SubscriptionData, Delta>(
-  query: Query<QueryData>,
-  subscriptionQuery: Query<SubscriptionData>,
-  update: Update<Data, Delta>,
-  getData: GetData<QueryData, Data>,
-  getDelta: GetDelta<SubscriptionData, Delta>,
-  pagination?: {
-    getPageInfo: GetPageInfo<QueryData>;
-    getTotalCount?: GetTotalCount<QueryData>;
-    append: Append<Data>;
-    first: number;
-  },
-  fetchPolicy: FetchPolicy = 'no-cache'
+  params: DataProviderParams<QueryData, Data, SubscriptionData, Delta>
 ): Subscribe<Data, Delta> {
   const getInstance = memoize<Data, Delta>(() =>
-    makeDataProviderInternal(
-      query,
-      subscriptionQuery,
-      update,
-      getData,
-      getDelta,
-      pagination,
-      fetchPolicy
-    )
+    makeDataProviderInternal(params)
+  );
+  return (callback, client, variables) =>
+    getInstance(variables)(callback, client, variables);
+}
+
+type DependencySubscribe = Subscribe<any, any>; // eslint-disable-line @typescript-eslint/no-explicit-any
+type DependencyUpdateCallback = Parameters<DependencySubscribe>['0'];
+export type CombineDerivedData<Data> = (
+  parts: Parameters<DependencyUpdateCallback>['0']['data'][]
+) => Data | null;
+
+function makeDerivedDataProviderInternal<Data>(
+  dependencies: DependencySubscribe[],
+  combineData: CombineDerivedData<Data>
+): Subscribe<Data, never> {
+  let subscriptions: ReturnType<DependencySubscribe>[] | undefined;
+  let client: ApolloClient<object> | undefined;
+  const callbacks: UpdateCallback<Data, never>[] = [];
+  let variables: OperationVariables | undefined;
+  const parts: Parameters<DependencyUpdateCallback>['0'][] = [];
+  let data: Data | null = null;
+  let error: Error | undefined;
+  let loading = true;
+  let loaded = false;
+
+  // notify single callback about current state, delta is passes optionally only if notify was invoked onNext
+  const notify = (callback: UpdateCallback<Data, never>) => {
+    callback({
+      data,
+      error,
+      loading,
+      loaded,
+      pageInfo: null,
+    });
+  };
+
+  // notify all callbacks
+  const notifyAll = () =>
+    callbacks.forEach((callback) => {
+      notify(callback);
+    });
+
+  const combine = () => {
+    let newError: Error | undefined;
+    let newLoading = false;
+    let newLoaded = true;
+    dependencies
+      .map((dependency, i) => parts[i])
+      .forEach((part) => {
+        newError = newError || (part && part.error);
+        newLoading = newLoading || !part || part.loading;
+        newLoaded = newLoaded && part && part.loaded;
+      });
+
+    const newData = newLoaded
+      ? combineData(parts.map((part) => part.data))
+      : data;
+    if (
+      newLoading !== loading ||
```

Comment:
suggestion (cosmetic): I'd write the conditions and assignments in the same order for better readability